### PR TITLE
Added missing e2e tests between Terra and Ethereum/Solana

### LINKED ISSUE #164: Complete Terra E2E tests
- 
  - [X] [Terra] Native -> [SOL] Wrapped
  - [x] [SOL] Wrapped -> [Terra] Native
- 
  - [X] [SOL] Native -> [Terra] Wrapped
  - [x] [Terra] Wrapped -> [SOL] Native
- 
  - [x] [Terra] Native -> [ETH] Wrapped
  - [x] [ETH] Wrapped -> [Terra] Native
- 
  - [x] [ETH] Native -> [Terra] Wrapped
  - [x] [Terra] Wrapped -> [ETH] Native

Note that the [SOL] Native -> [Terra] Wrapped test is unstable and fails ~50% of the time on a 19-node testnet.

### PR DESCRIPTION
**Pull request by ysavchenko**
Fixes #164 

This change is [Reviewable](https://reviewable.io/reviews/certusone/wormhole/168)

### CODE REVIEW

File: bridge/e2e/eth.go [hunk 1]
```diff
@@ -105,3 +106,78 @@ func testEthereumLockup(t *testing.T, ctx context.Context, ec *ethclient.Client,
 	// Source account decreases by the full amount.
 	waitEthBalance(t, ctx, token, beforeErc20, -int64(amount))
 }
+
+func testEthereumToTerraLockup(t *testing.T, ctx context.Context, ec *ethclient.Client, kt *bind.TransactOpts,
+	tc *TerraClient, tokenAddr common.Address, isNative bool, amount int64, precisionLoss int) {
```
Comment (leoluk) [inline bridge/e2e/eth.go:111]: Unused tc

File: bridge/e2e/eth.go [hunk 2]
```diff
@@ -105,3 +106,78 @@ func testEthereumLockup(t *testing.T, ctx context.Context, ec *ethclient.Client,
...
+	// Source token client
+	token, err := erc20.NewErc20(tokenAddr, ec)
+	if err != nil {
+		panic(err)
+	}
+
+	// Store balance of source ERC20 token
+	beforeErc20, err := token.BalanceOf(nil, devnet.GanacheClientDefaultAccountAddress)
+	if err != nil {
+		t.Log(err) // account may not yet exist, defaults to 0
```
Comment (leoluk) [inline bridge/e2e/eth.go:128]: Needs to be either t.Fatal or set `beforeErc20 = new(big.Int)` (this is wrong in the original function too - feel free to fix all occurrences)

File: bridge/e2e/eth.go [hunk 3]
```diff
@@ -105,3 +106,78 @@ func testEthereumLockup(t *testing.T, ctx context.Context, ec *ethclient.Client,
...
+	}
+	t.Logf("ERC20 balance: %v", beforeErc20)
+
+	// Store balance of destination CW20 token
+	paddedTokenAddress := padAddress(tokenAddr)
+	terraToken := devnet.TerraTokenAddress
+	if isNative {
+		terraToken, err = getAssetAddress(ctx, devnet.TerraBridgeAddress, vaa.ChainIDEthereum, paddedTokenAddress[:])
```
Comment (leoluk) [inline bridge/e2e/eth.go:136]: err not checked

File: bridge/e2e/solana.go [hunk 4]
```diff
@@ -120,13 +120,17 @@ func testSolanaLockup(t *testing.T, ctx context.Context, ec *ethclient.Client, c
...
+	sourceAcct string, tokenAddr string, isNative bool, amount int, precisionGain int) {
 
 	tokenSlice, err := base58.Decode(tokenAddr)
 	if err != nil {
 		t.Fatal(err)
 	}
-	terraToken, err := getAssetAddress(ctx, devnet.TerraBridgeAddress, vaa.ChainIDSolana, tokenSlice)
+	terraToken := devnet.TerraTokenAddress
+	if isNative {
+		terraToken = ""
```
Comment (leoluk) [inline bridge/e2e/solana.go:131]: What's this for?

File: bridge/e2e/terra.go [hunk 5]
```diff
@@ -166,9 +170,18 @@ func getAssetAddress(ctx context.Context, contract string, chain uint8, asset []
 	return gjson.Get(json, "result.address").String(), nil
 }
 
+func padAddress(address common.Address) vaa.Address {
```
Comment (leoluk) [inline bridge/e2e/terra.go:173]: We now have three copies of this, time to refactor into common package with some documentation what it does and replace all of them.

File: bridge/e2e/terra.go [hunk 6]
```diff
@@ -322,3 +335,54 @@ func testTerraLockup(t *testing.T, ctx context.Context, tc *TerraClient,
...
+	token, err := erc20.NewErc20(destination, ec)
+	if err != nil {
+		panic(err)
+	}
+
+	// Store balance of source CW20 token
+	beforeCw20, err := getTerraBalance(ctx, tokenAddr)
+	if err != nil {
+		t.Log(err) // account may not yet exist, defaults to 0
+		beforeCw20 = big.NewInt(0)
```
Comment (leoluk) [inline bridge/e2e/terra.go:351]: `new(big.Int)` for consistency

File: bridge/e2e/terra.go [hunk 7]
```diff
@@ -322,3 +335,54 @@ func testTerraLockup(t *testing.T, ctx context.Context, tc *TerraClient,
...
+	}
+	t.Logf("CW20 balance: %v", beforeCw20)
+
+	/// Store balance of wrapped destination token
+	beforeErc20, err := token.BalanceOf(nil, devnet.GanacheClientDefaultAccountAddress)
+	if err != nil {
+		t.Log(err) // account may not yet exist, defaults to 0
+		beforeErc20 = big.NewInt(0)
```
Comment (leoluk) [inline bridge/e2e/terra.go:359]: Same here

**Review by leoluk — CHANGES_REQUESTED** [review]

**Review by leoluk — CHANGES_REQUESTED** [review]
Also, the new tests consistently don't work on the first try:

```
=== RUN   TestEndToEnd_SOL_Terra/[SOL]_Native_->_[Terra]_Wrapped
    solana.go:139: cannot parse balance: 
    solana.go:141: CW20 balance: 0
2021/02/01 13:34:51 command: cli balance 3C3m4tjTy4nSMkkYdqCDSiCWEgpDa6whvprvABdFGBiW
    solana.go:148: SPL balance: 10000000000000000000
2021/02/01 13:34:52 command: cli lock Bridge1p5gheXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o 3C3m4tjTy4nSMkkYdqCDSiCWEgpDa6whvprvABdFGBiW 6qRhs8oAuZYLd4zzaNnQHqdRyknrQQWDWQhALEN8UA7M 50000000000 3 53639 00000000000000000000000035743074956c710800e83198011ccbd4ddf1556d
2021/02/01 13:34:53 command: cli balance 3C3m4tjTy4nSMkkYdqCDSiCWEgpDa6whvprvABdFGBiW
    solana.go:57: SPL balance after: 10000000000000000000 -> 9999999950000000000, delta -50000000000
    terra.go:291: cannot parse balance: 
    --- FAIL: TestEndToEnd_SOL_Terra/[SOL]_Native_->_[Terra]_Wrapped (8.08s)
```

```
=== RUN   TestEndToEnd_SOL_Terra/[Terra]_Wrapped_->_[SOL]_Native
    terra.go:301: cannot parse balance: 
    terra.go:303: CW20 balance: <nil>
2021/02/01 13:34:59 command: cli balance 3C3m4tjTy4nSMkkYdqCDSiCWEgpDa6whvprvABdFGBiW
    terra.go:310: SPL balance: 9999999950000000000
    terra.go:93: increaseAllowanceCall
         {"increase_allowance":{"spender":"terra174kgn5rtw4kf6f938wm7kwh70h2v4vcfd26jlc","amount":"50000000000"}}
    terra.go:94: lockAssetsCall
         {"lock_assets":{"asset":"","amount":"50000000000","recipient":"IIviHDdD3MJsl1Ixc17JwNQ6V0qyXaxonX5RJxZMxoc=","target_chain":1,"nonce":2523609398}}
    terra.go:327: non 200 respose code 500, error: {"error":"invalid address: missing contract: failed to simulate tx"}: failed to estimate fee [terra-project/terra.go@v1.0.1-0.20201113170042-b3bffdc6fd06/client/lcdclient.go:56]
    --- FAIL: TestEndToEnd_SOL_Terra/[Terra]_Wrapped_->_[SOL]_Native (0.23s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x18c2bae]

goroutine 662 [running]:
testing.tRunner.func1.1(0x1a6d8e0, 0x2bd8df0)
	/home/leopold/.local/lib/golang/go/src/testing/testing.go:1072 +0x30d
testing.tRunner.func1(0xc002075b00)
	/home/leopold/.local/lib/golang/go/src/testing/testing.go:1075 +0x41a
panic(0x1a6d8e0, 0x2bd8df0)
	/home/leopold/.local/lib/golang/go/src/runtime/panic.go:969 +0x1b9
github.com/certusone/wormhole/bridge/e2e.testTerraLockup(0xc002075b00, 0x20706a0, 0xc000126010, 0xc00131a850, 0xc000a21600, 0x0, 0x0, 0x1cf0af6, 0x2c, 0xba43b7400, ...)
	/home/leopold/projects/wormhole/bridge/e2e/terra.go:330 +0x46e
github.com/certusone/wormhole/bridge/e2e.TestEndToEnd_SOL_Terra.func4(0xc002075b00)
	/home/leopold/projects/wormhole/bridge/e2e/e2e_test.go:186 +0x245
testing.tRunner(0xc002075b00, 0xc001424a60)
	/home/leopold/.local/lib/golang/go/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/home/leopold/.local/lib/golang/go/src/testing/testing.go:1168 +0x2b3
```

**Review by ysavchenko — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/168)* status: 0 of 10 files reviewed, 9 unresolved discussions (waiting on @leoluk)

___
*[bridge/e2e/e2e_test.go, line 235 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1b-r1-235:-MSSfbd41oy_tdTwKsZb:bcizdyz) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/e2e_test.go#L235)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Here is it 12000000000 actually because of the precision difference.
___
*[bridge/e2e/e2e_test.go, line 256 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1b-r1-256:-MSSfTCAEZvyWqSi0RmE:b-t4v4zw) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/e2e_test.go#L256)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Isn't this just... 12? 
</blockquote></details>

Yes, basically it is, just the same value as in other ETH tests
___
*[bridge/e2e/eth.go, line 111 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1c-r1-111:-MSSgK4z5Ka1rSdCAtE5:b-896fix) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/eth.go#L111)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Unused tc
</blockquote></details>

Done.
___
*[bridge/e2e/eth.go, line 128 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1c-r1-128:-MSSgKZ22qdBWhkKHxYz:b-896fix) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/eth.go#L128)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Needs to be either t.Fatal or set `beforeErc20 = new(big.Int)` (this is wrong in the original function too - feel free to fix all occurrences)
</blockquote></details>

Done.
___
*[bridge/e2e/eth.go, line 136 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1c-r1-136:-MSShSN62HB-eaf2H2S2:b-896fix) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/eth.go#L136)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

err not checked
</blockquote></details>

Done.
___
*[bridge/e2e/solana.go, line 131 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1d-r1-131:-MSShTrID0WfRKfbKIZH:b-ccnea0) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/solana.go#L131)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

What's this for?
</blockquote></details>

It is a leftover from some other check, rewriting this whole part of the code.
___
*[bridge/e2e/terra.go, line 173 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1e-r1-173:-MSSoP066B9ftZzODptP:b-896fix) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/terra.go#L173)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

We now have three copies of this, time to refactor into common package with some documentation what it does and replace all of them.
</blockquote></details>

Done.
___
*[bridge/e2e/terra.go, line 351 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1e-r1-351:-MSSiw13C0R2J3AlSLHu:b-896fix) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/terra.go#L351)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

`new(big.Int)` for consistency
</blockquote></details>

Done.
___
*[bridge/e2e/terra.go, line 359 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1e-r1-359:-MSSiwUK3SpELRGym5QI:b-896fix) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/terra.go#L359)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Done.

**Review by leoluk — CHANGES_REQUESTED** [review]
```
# github.com/certusone/wormhole/bridge/e2e [github.com/certusone/wormhole/bridge/e2e.test]
./eth.go:135:24: undefined: "github.com/certusone/wormhole/bridge/pkg/ethereum".PadAddress
./terra.go:368:3: undefined: "github.com/certusone/wormhole/bridge/pkg/ethereum".PadAddress
./e2e_test.go:244:25: undefined: "github.com/certusone/wormhole/bridge/pkg/ethereum".PadAddress
```

I think you forgot to add that - please be more careful about proof-reading and testing commits :-) This just wastes the reviewer's time. I always self-review the change in the IDE before submitting it and would recommend you to do the same.

> Reviewed 7 of 7 files at r1, 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/168#-:-MST5BIaER8YKZpG4TXE:b-vhka8q)* status: all files reviewed, 5 unresolved discussions (waiting on @ysavchenko)

___
*[bridge/e2e/e2e_test.go, line 256 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1b-r1-256:-MST5lhq84pmXkB6rg1f:b-rr85qf) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/e2e_test.go#L256)):*
<details><summary><i>Previously, ysavchenko (Yuriy Savchenko) wrote…</i></summary><blockquote>

Yes, basically it is, just the same value as in other ETH tests
</blockquote></details>

Out of curiosity, why do we lose 9 digits of precision, given that: `TerraDefaultPrecision = 1e8`
___
*[bridge/e2e/e2e_test.go, line 68 at r2](https://reviewable.io/reviews/certusone/wormhole/168#-MST2vEcBn-LS4w5al7s:-MST2vEcBn-LS4w5al7t:bgv4t8q) ([raw file](https://github.com/certusone/wormhole/blob/bb4ce162ad3ea5d947bead7559a0a223e917b849/bridge/e2e/e2e_test.go#L68)):*
> ```Go
> // Either use different Ethereum account, or do not run Ethereum tests in parallel.
> 
> /* func TestEndToEnd_SOL_ETH(t *testing.T) {
> ```

Left-over debugging code
___
*[bridge/e2e/eth.go, line 140 at r2](https://reviewable.io/reviews/certusone/wormhole/168#-MST37b774dwjOMmxw5t:-MST37b774dwjOMmxw5u:bic821x) ([raw file](https://github.com/certusone/wormhole/blob/bb4ce162ad3ea5d947bead7559a0a223e917b849/bridge/e2e/eth.go#L140)):*
> ```Go
> 		terraToken, err = getAssetAddress(ctx, devnet.TerraBridgeAddress, vaa.ChainIDEthereum, paddedTokenAddress[:])
> 		if err != nil {
> 			t.Log(err)
> ```

Shouldn't this fail, or is it expected to continue with an empty `terraToken`?
___
*[bridge/e2e/solana.go, line 133 at r2](https://reviewable.io/reviews/certusone/wormhole/168#-MST3NE39kcB0AR17VSY:-MST3NE39kcB0AR17VSZ:b-xlp3nt) ([raw file](https://github.com/certusone/wormhole/blob/bb4ce162ad3ea5d947bead7559a0a223e917b849/bridge/e2e/solana.go#L133)):*
> ```Go
> 		terraToken, err = getAssetAddress(ctx, devnet.TerraBridgeAddress, vaa.ChainIDSolana, tokenSlice)
> 		if err != nil {
> 			t.Log(err)
> ```

same here, should it continue?
___
*[bridge/e2e/terra.go, line 214 at r2](https://reviewable.io/reviews/certusone/wormhole/168#-MST3TUa2JIideFMxTpH:-MST3TUa2JIideFMxTpI:b-4j9vv7) ([raw file](https://github.com/certusone/wormhole/blob/bb4ce162ad3ea5d947bead7559a0a223e917b849/bridge/e2e/terra.go#L214)):*
> ```Go
> 		}
> 
> 		if address == "" {
> ```

Please add a comment explaining this case

**Review by ysavchenko — COMMENTED** [review]
Sorry, the import itself is correct, but forgot to add a new file with this function. I know how frustrating it could be, fixing this now.

> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/168#-:-MST7wP307zOfdMlY8BO:bfce40b)* status: all files reviewed, 5 unresolved discussions (waiting on @ysavchenko)

**Review by ysavchenko — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/168)* status: 8 of 11 files reviewed, 5 unresolved discussions (waiting on @leoluk)

___
*[bridge/e2e/e2e_test.go, line 256 at r1](https://reviewable.io/reviews/certusone/wormhole/168#-MSRxd_gJ3pJf0lliK1b-r1-256:-MST8I69Cfvst9XRi6PQ:bmo355) ([raw file](https://github.com/certusone/wormhole/blob/cb0bbc4430bec5a6a777d3cd97077b8a10b3e1d3/bridge/e2e/e2e_test.go#L256)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Out of curiosity, why do we lose 9 digits of precision, given that: `TerraDefaultPrecision = 1e8`
</blockquote></details>

I believe this is the logic within the VAA processing to cap precision at 9 digits. It is ok for Terra (because 9 digits work fine) and 8 digits is just a default precision there.
___
*[bridge/e2e/e2e_test.go, line 68 at r2](https://reviewable.io/reviews/certusone/wormhole/168#-MST2vEcBn-LS4w5al7s:-MST9mhU1tAyhbpEz_II:b-896fix) ([raw file](https://github.com/certusone/wormhole/blob/bb4ce162ad3ea5d947bead7559a0a223e917b849/bridge/e2e/e2e_test.go#L68)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Left-over debugging code
</blockquote></details>

Done.
___
*[bridge/e2e/eth.go, line 140 at r2](https://reviewable.io/reviews/certusone/wormhole/168#-MST37b774dwjOMmxw5t:-MST8mWj3JmbMtB8PiNR:bpdnp59) ([raw file](https://github.com/certusone/wormhole/blob/bb4ce162ad3ea5d947bead7559a0a223e917b849/bridge/e2e/eth.go#L140)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Shouldn't this fail, or is it expected to continue with an empty `terraToken`?
</blockquote></details>

Yes, this could continue. If it is empty then we consider the initial balance to be 0 and try to retrieve the token address later.
___
*[bridge/e2e/solana.go, line 133 at r2](https://reviewable.io/reviews/certusone/wormhole/168#-MST3NE39kcB0AR17VSY:-MST8v_M1yqpmQ6WUJdS:bmvd1lw) ([raw file](https://github.com/certusone/wormhole/blob/bb4ce162ad3ea5d947bead7559a0a223e917b849/bridge/e2e/solana.go#L133)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

same here, should it continue?
</blockquote></details>

Same :)
___
*[bridge/e2e/terra.go, line 214 at r2](https://reviewable.io/reviews/certusone/wormhole/168#-MST3TUa2JIideFMxTpH:-MST9ZCF5dNT7T04U8q9:b1kkkwb) ([raw file](https://github.com/certusone/wormhole/blob/bb4ce162ad3ea5d947bead7559a0a223e917b849/bridge/e2e/terra.go#L214)):*
<details><summary><i>Previously, leoluk (Leopold Schabel) wrote…</i></summary><blockquote>

Please add a comment explaining this case
</blockquote></details>

It is basically for the case when request for asset address is successful (so there is no error), but address itself is not in the registry. Adding this as a comment.

**Review by leoluk — APPROVED** [review]
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/168)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**